gpio/gpiotest/gpiotest.go

```diff
@@ -12,11 +12,14 @@ import (
 	"sync"
 	"time"
 
+	"github.com/jonboulle/clockwork"
 	"periph.io/x/conn/v3/gpio"
 	"periph.io/x/conn/v3/physic"
 	"periph.io/x/conn/v3/pin"
 )
 
+var Clock = clockwork.NewRealClock()
```

Comment:
Instead, do:

```
var (
  after = time.After
  now = time.Now
)
```

In particular, we don't want it exported. Then:

- replace the calls with these two function variables. Replace time.Since() with now().Sub().
- replace them in the test similar to what you did.

I really prefer conn/ to stay a pure Go package unless absolutely necessary and I don't feel it pass the bar. Sorry for the trouble.

Reply:
I understand the need to keep it pure (presumably you mean dependency-free; the clockwork dependency is still pure Go, I think), and it's no trouble. However, this doesn't quite work:

* If `gpiotest.after` and `gpiotest.now` are unexported, the test (which is in package `gpioutil`) can't modify them
* Even if they are exported and the test works, this still requires adding `clockwork` to `go.mod`, although it is now marked `// indirect`. Is that sufficient decoupling?

Reply:
Ohhh sorry I botched my review. Since it's gpiotest, make it a variable but inside Pin. Then Now and After can be per pin where nil is a valid default, and it's easy for both implementation and call sites.

For debounced, either do the same, make a global or hack by type cast testing (*gpiotest.Pin). I don't have a preference yet.

Does that sound reasonable?

Reply:
Hmm, this uncovers something interesting; `gpiotest.LogPinIO.WaitForEdge` doesn't have access to those variables, as it gets a `LogPinIO`. Come to think of it, I guess it makes more sense for it to use the real clock? And then we don't even need to fake `time.Now`?

Also, I wonder if the syntax I'm using is reasonable. It becomes (in `TestDebounce_WaitForEdge_Noise_NoEdge`):

```go
p.(*debounced).PinIO.(*gpiotest.Pin).After = fakeClock.After
defer func() {
	p.(*debounced).PinIO.(*gpiotest.Pin).After = nil
}()
```

Finally, this still leaves the issue of `go.mod`; WDYT?

Reply:
There's no reason for the defer. It's a temporary object.

I don't mind not "fixing" LogPinIO, since the odds of it being used with a gpiotest.Pin underneath is frankly low, I'm not sure what use case would make sense that would also mock the time. Especially that in this case time is only used for output. So I feel that real time being used is a feature, not a bug. Hence WAI.

Then I don't feel like there's a need for the package if you reimplement minimal aspects of it locally?

Reply:
> There's no reason for the defer. It's a temporary object.
> I don't mind not "fixing" LogPinIO, since the odds of it being used with a gpiotest.Pin underneath is frankly low, I'm not sure what use case would make sense that would also mock the time. Especially that in this case time is only used for output. So I feel that real time being used is a feature, not a bug. Hence WAI.

Yep, fair on both counts.

> Then I don't feel like there's a need for the package if you reimplement minimal aspects of it locally?

It's not that minimal; while `gpiotest.go` no longer references it, `debounce.go` does. `WaitForEdge` already uses `clockwork`'s `Clock.Now`, `Clock.Since` and `Clock.Sleep`, in additionl to `debounce_test.go` using its `FakeClock.BlockUntil` and `FakeClock.Advance`; the implementations of those two in particular are nontrivial and subtle (making me glad someone else implemented it and that it's widely-ish-used). Options include:

* Accept this dependency
* Reject this dependency outright, go back to actually sleeping in tests (longer and flakier as a result)
* Copy substantial amounts of code from `clockwork` into `conn`
* Modify `debounce.go` so that it doesn't reference `clockwork`, but `debounce_test.go` still does (but I don't understand the benefit to this, as `go.mod` will still reference `clockwork`)
* Find a cleverer, simpler way to test this so that `clockwork` is not necessary (I have failed at this)

WDYT?

Reply:
Oops, time flies. I did my grief, we can depend on it. :)

Still, I'd prefer this to not be a global, since it will make test brittle and not able to run in parallel.

So it should be a member of Pin.